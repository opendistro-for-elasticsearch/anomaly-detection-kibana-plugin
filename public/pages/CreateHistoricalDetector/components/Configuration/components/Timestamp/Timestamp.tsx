/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiComboBox, EuiCallOut, EuiSpacer, EuiFlexItem } from '@elastic/eui';
import { Field, FieldProps, FormikProps } from 'formik';
import { debounce, get, isEmpty } from 'lodash';
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ContentPanel from '../../../../../../components/ContentPanel/ContentPanel';
import { AppState } from '../../../../../../redux/reducers';
import { getPrioritizedIndices } from '../../../../../../redux/reducers/elasticsearch';
import { FormattedFormRow } from '../../../../../createDetector/components/FormattedFormRow/FormattedFormRow';
import { sanitizeSearchText } from '../../../../../utils/helpers';
import { getError, isInvalid, required } from '../../../../../../utils/utils';
import { HistoricalDetectorFormikValues } from '../../../../utils/constants';

interface TimestampProps {
  formikProps: FormikProps<HistoricalDetectorFormikValues>;
}

export function Timestamp(props: TimestampProps) {
  const dispatch = useDispatch();
  const elasticsearchState = useSelector(
    (state: AppState) => state.elasticsearch
  );
  const selectedIndex = get(props.formikProps, 'values.index.0.label', '');
  const isIndexSelected = selectedIndex && selectedIndex.length > 0;
  const isRemoteIndex = selectedIndex.includes(':');

  const [queryText, setQueryText] = useState('');

  const handleSearchChange = debounce(async (searchValue: string) => {
    if (searchValue !== queryText) {
      const sanitizedQuery = sanitizeSearchText(searchValue);
      setQueryText(sanitizedQuery);
      await dispatch(getPrioritizedIndices(sanitizedQuery));
    }
  }, 300);

  const dateFields = Array.from(
    get(elasticsearchState, 'dataTypes.date', []) as string[]
  );

  const timeStampFieldOptions = isEmpty(dateFields)
    ? []
    : dateFields.map((dateField) => ({ label: dateField }));

  return (
    <ContentPanel title="Timestamp" titleSize="s">
      {isRemoteIndex ? (
        <div>
          <EuiCallOut
            title="A remote index is selected, so you need to manually input the time field."
            color="warning"
            iconType="alert"
          />
          <EuiSpacer size="m" />
        </div>
      ) : null}
      <Field name="timeField" validate={required}>
        {({ field, form }: FieldProps) => (
          <EuiFlexItem style={{ maxWidth: '70%' }}>
            <FormattedFormRow
              title="Timestamp field"
              hint="Choose the time field you want to use for time filter."
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
            >
              <EuiComboBox
                id="timeField"
                placeholder="Find timestamp"
                options={isIndexSelected ? timeStampFieldOptions : []}
                onSearchChange={handleSearchChange}
                onCreateOption={(createdOption: string) => {
                  const normalizedOptions = createdOption.trim();
                  if (!normalizedOptions) return;
                  form.setFieldValue('timeField', normalizedOptions);
                }}
                onBlur={() => {
                  form.setFieldTouched('timeField', true);
                }}
                onChange={(options) => {
                  form.setFieldValue('timeField', get(options, '0.label'));
                }}
                selectedOptions={
                  (field.value && [{ label: field.value }]) || []
                }
                singleSelection={true}
                isClearable={true}
              />
            </FormattedFormRow>
          </EuiFlexItem>
        )}
      </Field>
    </ContentPanel>
  );
}
