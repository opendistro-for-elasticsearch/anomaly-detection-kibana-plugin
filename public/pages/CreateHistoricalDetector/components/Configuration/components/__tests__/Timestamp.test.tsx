/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render, fireEvent, wait } from '@testing-library/react';
import { Timestamp } from '../Timestamp/Timestamp';
import configureStore from '../../../../../../redux/configureStore';
import {
  httpClientMock,
  coreServicesMock,
} from '../../../../../../../test/mocks';
import { CoreServicesContext } from '../../../../../../components/CoreServices/CoreServices';
import { Formik } from 'formik';
import userEvent from '@testing-library/user-event';

const TITLE_TEXT = 'Timestamp';
const CALLOUT_TEXT =
  'A remote index is selected, so you need to manually input the time field.';
const INVALID_TEXT = 'Required';

const REGULAR_INDEX = [{ label: 'regular-index' }];
const REMOTE_INDEX = [{ label: 'remote:index' }];

const renderWithRouter = (indexArray: [] = []) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <Formik
                  initialValues={{ index: indexArray }}
                  onSubmit={jest.fn()}
                >
                  {() => (
                    <div>
                      <Timestamp
                        formikProps={{
                          values: { index: indexArray },
                          setFieldValue: jest.fn(),
                          setFieldTouched: jest.fn(),
                        }}
                        {...props}
                      />
                    </div>
                  )}
                </Formik>
              </CoreServicesContext.Provider>
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<Timestamp /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders the component with regular index', () => {
    const { container, getByText, queryByText } = renderWithRouter(
      REGULAR_INDEX
    );
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    expect(queryByText(CALLOUT_TEXT)).toBeNull();
  });
  test('renders the component with remote index', () => {
    const { container, getByText } = renderWithRouter(REMOTE_INDEX);
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    getByText(CALLOUT_TEXT);
  });
  test('catches invalid empty timestamp', async () => {
    const { getByText } = renderWithRouter();
    userEvent.click(getByText('Find timestamp'));
    fireEvent.blur(getByText('Find timestamp'));
    await wait();
    getByText(TITLE_TEXT);
    getByText(INVALID_TEXT);
  });
});
