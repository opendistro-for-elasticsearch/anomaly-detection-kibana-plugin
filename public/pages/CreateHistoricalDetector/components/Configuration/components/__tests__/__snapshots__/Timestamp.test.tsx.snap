// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timestamp /> spec renders the component with regular index 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Timestamp
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFlexItem"
        style="max-width: 70%;"
      >
        <div
          class="euiFormRow"
          hint="Choose the time field you want to use for time filter."
          id="random_html_id-row"
          title="Timestamp field"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Timestamp field
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Choose the time field you want to use for time filter.
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              class="euiComboBox"
              role="combobox"
            >
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <div
                    class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                    data-test-subj="comboBoxInput"
                    tabindex="-1"
                  >
                    <p
                      class="euiComboBoxPlaceholder"
                    >
                      Find timestamp
                    </p>
                    <div
                      class="euiComboBox__input"
                      style="font-size: 14px; display: inline-block;"
                    >
                      <input
                        aria-controls=""
                        data-test-subj="comboBoxSearchInput"
                        id="random_html_id"
                        role="textbox"
                        style="box-sizing: content-box; width: 2px;"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                  <div
                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <button
                      aria-label="Open list of options"
                      class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                      data-test-subj="comboBoxToggleListButton"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                        focusable="false"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Timestamp /> spec renders the component with remote index 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Timestamp
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div>
        <div
          class="euiCallOut euiCallOut--warning"
        >
          <div
            class="euiCallOutHeader"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiCallOutHeader__icon"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
            <span
              class="euiCallOutHeader__title"
            >
              A remote index is selected, so you need to manually input the time field.
            </span>
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--m"
        />
      </div>
      <div
        class="euiFlexItem"
        style="max-width: 70%;"
      >
        <div
          class="euiFormRow"
          hint="Choose the time field you want to use for time filter."
          id="random_html_id-row"
          title="Timestamp field"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Timestamp field
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Choose the time field you want to use for time filter.
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              class="euiComboBox"
              role="combobox"
            >
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <div
                    class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                    data-test-subj="comboBoxInput"
                    tabindex="-1"
                  >
                    <p
                      class="euiComboBoxPlaceholder"
                    >
                      Find timestamp
                    </p>
                    <div
                      class="euiComboBox__input"
                      style="font-size: 14px; display: inline-block;"
                    >
                      <input
                        aria-controls=""
                        data-test-subj="comboBoxSearchInput"
                        id="random_html_id"
                        role="textbox"
                        style="box-sizing: content-box; width: 2px;"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                  <div
                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <button
                      aria-label="Open list of options"
                      class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                      data-test-subj="comboBoxToggleListButton"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                        focusable="false"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                          fill-rule="non-zero"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
